ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/transaction/arjunacore/jca/XATerminatorImple.java

```diff
@@ -241,6 +241,12 @@ public int prepare (Xid xid) throws XAException
 
             switch (tx.doPrepare())
             {
+            case TwoPhaseOutcome.PREPARE_ONE_PHASE_COMMITTED:
+                // Should never happen Would be great if there was heuristic commit :(
+                SubordinationManager.getTransactionImporter()
+                        .removeImportedTransaction(xid);
+                jtaLogger.i18NLogger.fatalSubordinate1PCDuringPrepare(xid);
+                throw new XAException(XAException.XAER_RMERR);
```

Comment:
Unfortunately RMERR is not perfect either at it suggests it couldn't prepare rather than a heuristic commit of the second resource. With my patch this outcome would hopefully never be seen anyway.